### DISCUSSION TITLE
How to derive system model Jacobian for $SE_2(3)$?

### DISCUSSION THREAD

**Discussion by Chris7462** [artivis/manif#309] [Q&A] [answered]
I have been exploring the $SE_2(3)$ localization example and am curious about how the system model Jacobian is calculated, specifically how to compute $J^{u}_{x}$?

In the code example
https://github.com/artivis/manif/blob/de52f8b8bf354e5e03a61025986f4fdb1abad774/examples/se_2_3_localization.cpp#L376-L383

I understand that:

-  $J^{x}_{x}=Ad\_{Exp(\tau)}^{-1}$, as mentioned in Eq. (80) from manif paper
-  $J^{x}\_{u}=J_{r}(\tau)$, as per Eq. (81)

but I'm unsure about how to derive $J^{u}_{x}$.

Additionally, I've reviewd [Barrau and Bonnabel's papaer](http://arxiv.org/pdf/1410.1465), but it hasn't provided the clarity I was hoping for.

**Comment by joansola** [comment]
On lline 371 in the cpp file you have the expression of `u_est`. On line 376 you have `X <- X (+) u_est`, with Jacobians `J_X_X` and `J_X_u`.

It happens in this case that `u_est` is itself a function of `X`. Therefore, there exists a `J_u_X` which has to be computed and then used in the chain rule, line 383.

Now your question is about `J_u_X`. For this, you look at the code to copute `u_est` from `X`, and work out the derivative. This code is in lines 362--369, just above `u_est`.

The derivatives of these mathematical expressions I guess shoulde not be difficult. Work lines 362-369 line by line, using the formulas in the micro Lie paper, and then apply the chain rule to get `J_u_X`.

For example, if a line states `accLin = dt*(R_k_est.transpose())*v_k_est + 0.5*dt*dt*acc_k_est`, then compute the Jacobians: `J_acclin_R, J_acciln_v, J_acclin_acc`. They should be most quite easy. Look at the derivation rules in the paper, and in the Appendix related to SE(3) -- the Jacobians of SE_2(3) are block-wise equivalent to those of SE(3).

Once you have all lines, and the good names for all partial Jacobians, then apply the chain rule. Good names are such that if `z=f(x,y)` then you name `J_z_x = df/dx` and `J_z_y=df/dy`. That is, `J_<output variablee name>_<input variable name>`. Use good names otherwise the chain rule is impossible to get right.

Is this response enough?

**Comment by joansola** [comment, marked as answer]
Markiing previous answer as valid...
